tests/types/test_numpy.py

```diff
@@ -0,0 +1,376 @@
+import pytest
+from psycopg.adapt import PyFormat
+from psycopg.types.numpy import register_numpy
+
+pytest.importorskip("numpy")
+
+import numpy as np  # noqa: E402
+
+pytestmark = [pytest.mark.numpy]
+
+# Numpy numerical types
+#
+# np.byte  (-128, 127) -> smallint
+# np.short (-32768, 32767) -> smallint
+# np.intc  (-2147483648, 2147483647) -> integer
+# np.int_ (-9223372036854775808, 9223372036854775807) -> bigint (alias int8)
+# np.ubyte (0, 255) -> smallint
+# np.ushort (0, 65535) -> integer
+# np.uintc (0, 4294967295) -> bigint
+# np.uint (0, 18446744073709551615) -> numeric ?
+
+
+@pytest.fixture
+def numpy_conn(conn):
+    register_numpy(conn)
+    return conn
+
+
+@pytest.mark.parametrize(
+    "val, expr",
+    [
+        (
+            np.byte(-128),
+            "'-128'::int2",
+        ),
+        (
+            np.byte(127),
+            "'127'::int2",
+        ),
+        (
+            np.byte(0),
+            "'0'::int2",
+        ),
+        (
+            np.byte(45),
+            "'45'::int2",
+        ),
+        # byte overflow
+        (
+            np.byte(-129),
+            "'127'::int2",
+        ),
+        (
+            np.byte(128),
+            "'-128'::int2",
+        ),
+    ],
+)
+@pytest.mark.parametrize("fmt_in", PyFormat)
+def test_dump_numpy_byte(numpy_conn, val, expr, fmt_in):
+
+    assert isinstance(val, np.byte)
+    assert np.byte is np.int8
+
+    cur = numpy_conn.cursor()
+
+    cur.execute(f"select pg_typeof({expr}) = pg_typeof(%{fmt_in.value})", (val,))
+    cur.fetchone()[0] is True
+
+    cur.execute(f"select {expr} = %{fmt_in.value}", (val,))
+    assert cur.fetchone()[0] is True
+
+
+@pytest.mark.parametrize(
+    "val, expr",
+    [
+        (
+            np.short(-32_768),
+            "'-32768'::int2",
+        ),
+        (
+            np.short(32_767),
+            "'32767'::int2",
+        ),
+        (
+            np.short(0),
+            "'0'::int2",
+        ),
+        (
+            np.short(45),
+            "'45'::int2",
+        ),
+        # short overflow
```

Comment:
I was very confused by this behavior as well, and not sure what I had to test in terms of overflow. I read that this “silent wrap around” is not guaranteed for the future, so I added that test to make sure that the behavior will remain consistent.

Reply:
Even if the behaviour changes, I don't think there's much we can do. The test will fail, but not for psycopg behaviour or misbehaviour. A short int will surely not contain a value > 32767; the only behaviour I see changing from their part is that the constructor may overflow instead of clipping the value to max.